GigE: Support GigE cameras and PTP timesync

jackg0 opened the pull request:
Summary: Enabling GigE support and PTP timesync.

Test plan: Testing daily with local bench setup.

jackg0 committed:
GigE: Patch issue with U3VMessageChannelID on GigE (https://github.com/ros-drivers/flir_camera_driver/issues/27)

jackg0 committed:
GigE: Change incomplete image error to warning and discard image.

jackg0 committed:
GigE: Support color streaming.

jackg0 committed:
PTP: Support enabling ptp with dynamic reconfigure.

jackg0 committed:
GigE: Find camera properly when multiple GigE devices are available.

Commit:
nodelet: add option to publish images over iceoryx transport.

Commit:
wip: support iox publisher.

Commit:
build iox transport support into oryx nodelet.

Commit:
don't free sent chunks.

jackg0 committed:
wip: Fix timestamping bug in image header.

ziparker commented on spinnaker_camera_driver/CMakeLists.txt:
```diff
@@ -108,10 +107,6 @@ if(EXISTS ${CMAKE_CURRENT_BINARY_DIR}/usr/lib)
            PATTERN libQt* EXCLUDE)
 else()
   get_filename_component(Spinnaker_LIB_DIR ${Spinnaker_LIBRARIES} DIRECTORY)
-  file(READ ${PROJECT_SOURCE_DIR}/cmake/install_files filenames)
```
can you explain the issue behind this removal?
just want to be careful about changes in case we end up with something to contribute upstream.

jackg0 replied:
it was breaking on `libSpinnaker.so.1` in `install_files` because we have `libSpinnaker.so.2` installed. We install spinnaker outside of the build process of this package, so it didn't seem right to try to install it again here.

ziparker commented on spinnaker_camera_driver/cfg/Spinnaker.cfg:
```diff
@@ -111,6 +111,9 @@ gen.add("saturation",                            double_t,    SensorLevels.RECON
 gen.add("gamma",                                 double_t,    SensorLevels.RECONFIGURE_RUNNING,     "Controls the gamma correction of pixel intensity.",                                                  1.0,                          0.5,   4.0)
 
+# PTP
+gen.add("enable_ptp",                     bool_t,      SensorLevels.RECONFIGURE_RUNNING,     "Enables/disables Saturation adjustment..",                                                           True)
```
update description.

ziparker commented on spinnaker_camera_driver/launch/camera.launch:
```diff
@@ -46,19 +61,35 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
     <node pkg="nodelet" type="nodelet" name="image_proc_debayer"
           args="load image_proc/debayer camera_nodelet_manager">
     </node>
+
```
remove whitespace if it was consistent before

ziparker commented on spinnaker_camera_driver/src/SpinnakerCamera.cpp:
```diff
@@ -337,14 +346,16 @@ void SpinnakerCamera::grabImage(sensor_msgs::Image* image, const std::string& fr
-        image->header.stamp.sec = image_ptr->GetTimeStamp() * 1e-9;
-        image->header.stamp.nsec = image_ptr->GetTimeStamp();
+        ros::Time imgStamp(image_ptr->GetTimeStamp() * 1e-9);
+        ros::Duration ptpOffset(-37);
```
comment on the -37 (tai/utc offset is not accounted for in sensor output, apparently).

ziparker commented on spinnaker_camera_driver/src/nodelet.cpp:
```diff
@@ -586,10 +597,12 @@ class SpinnakerCameraNodelet : public nodelet::Nodelet
+            if (!enable_ptp_)
+            {
+              ros::Time time = ros::Time::now();
+              wfov_image->header.stamp = time;
```
if ptp is enabled, should we still set the stamp field in the image header?

ziparker commented on spinnaker_camera_driver/src/nodelet.cpp:
```diff
@@ -705,6 +718,8 @@ class SpinnakerCameraNodelet : public nodelet::Nodelet
   bool do_rectify_;  ///< Whether or not to rectify as if part of an image.  Set to false if whole image, and true if in
                      /// ROI mode.
 
+  bool enable_ptp_;
```
might have missed this when scrolling through - make sure this is always init'd by the ctor(s).
i'd normally say init here, but looks like that isn't the convention in this code.

jackg0 replied:
it's currently initialized on this line: https://github.com/jackg0/flir_camera_driver/blob/gigevision-support/spinnaker_camera_driver/src/nodelet.cpp#L141
I could go through and change the way the drivier inits these variables if you want.

jackg0 committed:
wip: Fix PTP dynamic cfg setting description.

jackg0 committed:
wip: Fix PR comments.

Commit:
fix build issue from gige branch.

Commit:
use iox pub wrapper.

Commit:
match iox & ros topics.

jackg0 committed:
wip: Fix launch file.

jackg0 committed:
wip: Fix indentation.

jackg0 committed:
Merge pull request #2 from ziparker/ioxpub

add iox publisher support to the camera nodelet.
this also changes connection callback usage to poll for subscriptions to avoid potential race conditions and allow configuration to publish to the debayer nodelet (or elsewhere) unconditionally.

jackg0 committed:
wip: Check-in image header update.

jackg0 committed:
wip: Update ptp timestamping.

jackg0 committed:
wip: Support setting Target Grey Value.

jackg0 committed:
wip: Set target grey value properties in node map correctly.

jackg0 committed:
wip: Fix comments in cfg file.

Commit:
nodelet: fix segfault in devicePoll().

device poll can be called prior to pub_ instantiation, since the connection
callback is created prior to the publisher ctor in onInit().

null-check the publisher in the connection callback to avoid triggering
segfaults when the race condition is triggered.

note, this doesn't fix the race condition, since that'd require messing with
the connection callback setup, which is required for creating the publisher,
but we shouldn't need to resolve that in any other way.

this normally happened when starting a bunch of other ros nodes along with the
oryx node.

jackg0 committed:
Merge pull request #3 from ziparker/gigevision-support

nodelet: fix segfault in devicePoll().

Commit:
wip: Add new launch file without arguments.

jackg0 committed:
wip: Add params to launch file.

Commit:
wip: Add params to launch file.

jackg0 committed:
Cleanup diagnostics.

jackg0 committed:
Improve and add diagnostics.

jackg0 committed:
Fix debug message for ptp status.

jackg0 committed:
Catch errors by const reference.
